hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/CallIdSequenceWithBackoff.java

```diff
@@ -0,0 +1,80 @@
+package com.hazelcast.client.spi.impl;
+
+import com.hazelcast.core.HazelcastOverloadException;
+import com.hazelcast.util.concurrent.BackoffIdleStrategy;
+import com.hazelcast.util.concurrent.IdleStrategy;
+
+import java.util.concurrent.atomic.AtomicLongArray;
+
+import static com.hazelcast.nio.Bits.CACHE_LINE_LENGTH;
+import static com.hazelcast.nio.Bits.LONG_SIZE_IN_BYTES;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+/**
+ * A {@link com.hazelcast.spi.impl.operationservice.impl.CallIdSequence} that provided backpressure by taking
+ * the number in flight operations into account when a call-id needs to be determined.
+ *
+ * It is possible to temporary exceed the capacity:
+ * - due to system operations
+ * - due to racy nature of checking if space is available and getting the next sequence.
+ */
+public final class CallIdSequenceWithBackoff extends CallIdSequence {
+    private static final int MAX_DELAY_MS = 500;
+    private static final IdleStrategy IDLER = new BackoffIdleStrategy(
+            0, 0, MILLISECONDS.toNanos(1), MILLISECONDS.toNanos(MAX_DELAY_MS));
+
+    private static final int INDEX_HEAD = 7;
+    private static final int INDEX_TAIL = INDEX_HEAD + CACHE_LINE_LENGTH / LONG_SIZE_IN_BYTES;
+
+    // instead of using two AtomicLongs, we use an array of width of 3 cache lines to prevent any false sharing
+    private final AtomicLongArray longs = new AtomicLongArray(3 * CACHE_LINE_LENGTH / LONG_SIZE_IN_BYTES);
+
+    private final int maxConcurrentInvocations;
+    private final long backoffTimeoutMs;
+
+    public CallIdSequenceWithBackoff(int maxConcurrentInvocations, long backofftimeoutMs) {
+        this.maxConcurrentInvocations = maxConcurrentInvocations;
+        this.backoffTimeoutMs = backofftimeoutMs;
+    }
+
+    @Override
+    public long next() {
+        if (!hasSpace()) {
+            waitForSpace();
+        }
+
+        return longs.incrementAndGet(INDEX_HEAD);
+    }
+
+    private void waitForSpace() {
+        if (backoffTimeoutMs <= 0) {
```

Comment:
This is not exactly OOP code, but it's simple and perhaps easier to read then class hierarchies and what not.
I am open for suggestions tho.